src/util/string.h

```diff
@@ -17,6 +17,29 @@
 #include <vector>
 
 namespace util {
+/** Type to denote a format string that was checked at compile time */
+template <int num_params>
+struct ConstevalFormatString {
+    const char* const fmt;
+    consteval ConstevalFormatString(const char* str) : fmt{str} { Detail_CheckNumFormatSpecifiers(fmt); }
+    constexpr static void Detail_CheckNumFormatSpecifiers(std::string_view str)
+    {
+        int count{0};
```

Comment:
If you keep this version, `num_params` can likely be a `size_t` instead:
```diff
diff --git a/src/util/string.h b/src/util/string.h
--- a/src/util/string.h	(revision fa72ce66421d3f90a6794b3e54e56873ae81265f)
+++ b/src/util/string.h	(date 1725561949080)
@@ -18,13 +18,13 @@
 
 namespace util {
 /** Type to denote a format string that was checked at compile time */
-template <int num_params>
+template <size_t num_params>
 struct ConstevalFormatString {
     const char* const fmt;
     consteval ConstevalFormatString(const char* str) : fmt{str} { Detail_CheckNumFormatSpecifiers(fmt); }
     constexpr static void Detail_CheckNumFormatSpecifiers(std::string_view str)
     {
-        int count{0};
+        size_t count{0};
         bool begin_f{false};
         for (char c : str) {
             bool perc{c == '%'};
```

Reply:
it's not the counting, it's to document that this version doesn't have negatives (like the excess version does)